File: app/src/main/java/com/ssy/ferry/core/FerryHandlerThread.java
package com.ssy.ferry.core;

import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;

import java.util.HashSet;
import java.util.Iterator;

/**
 * Created by zhangshaowen on 17/7/5.
 */

public class FerryHandlerThread {
    private static final String TAG = "Ferry.HandlerThread";

    public static final String Ferry_THREAD_NAME = "default_Ferry_thread";

    /**
     * unite defaultHandlerThread for lightweight work,
     * if you have heavy work checking, you can create a new thread
     */
    private static volatile HandlerThread defaultHandlerThread;
    private static volatile Handler defaultHandler;
    private static volatile Handler defaultMainHandler = new Handler(Looper.getMainLooper());
    private static HashSet<HandlerThread> handlerThreads = new HashSet<>();

    public static Handler getDefaultMainHandler() {
        return defaultMainHandler;
    }

    public static HandlerThread getDefaultHandlerThread() {

        synchronized (FerryHandlerThread.class) {
            if (null == defaultHandlerThread) {
                defaultHandlerThread = new HandlerThread(Ferry_THREAD_NAME);
                defaultHandlerThread.start();
                defaultHandler = new Handler(defaultHandlerThread.getLooper());
            }
            return defaultHandlerThread;
        }
    }

    public static Handler getDefaultHandler() {
        return defaultHandler;
    }

    public static HandlerThread getNewHandlerThread(String name) {
        for (Iterator<HandlerThread> i = handlerThreads.iterator(); i.hasNext();) {
            HandlerThread element = i.next();
            if (!element.isAlive()) {
                i.remove();
            }
        }
        HandlerThread handlerThread = new HandlerThread(name);
        handlerThread.start();
        handlerThreads.add(handlerThread);
        return handlerThread;
    }
}

File: app/src/main/java/com/ssy/ferry/core/FerryHandlerThread.kt
package com.ssy.ferry.core

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import java.util.HashSet

/**
 * 2019-10-17
 * @author Mr.S
 */
class FerryHandlerThread {


    companion object {
        private val TAG = "Ferry.HandlerThread"
        val Ferry_THREAD_NAME = "default_Ferry_thread"
        /**
         * unite defaultHandlerThread for lightweight work,
         * if you have heavy work checking, you can create a new thread
         */
        private var defaultHandlerThread: HandlerThread? = null
        private var defaultHandler: Handler? = null
        private val defaultMainHandler = Handler(Looper.getMainLooper())
        private val handlerThreads = HashSet<HandlerThread>()

        fun getDefaultMainHandler(): Handler {
            return defaultMainHandler
        }

        fun getDefaultHandlerThread(): HandlerThread? {

            synchronized(FerryHandlerThread::class.java) {
                if (null == defaultHandlerThread) {
                    defaultHandlerThread = HandlerThread(Ferry_THREAD_NAME)
                    defaultHandlerThread!!.start()
                    defaultHandler = Handler(defaultHandlerThread!!.looper)
                }
                return defaultHandlerThread
            }
        }

        fun getDefaultHandler(): Handler? {
            return defaultHandler
        }

        fun getNewHandlerThread(name: String): HandlerThread {
            val i = handlerThreads.iterator()
            while (i.hasNext()) {
                val element = i.next()
                if (!element.isAlive) {
                    i.remove()
                }
            }
            val handlerThread = HandlerThread(name)
            handlerThread.start()
            handlerThreads.add(handlerThread)
            return handlerThread
        }
    }

}